pkg/server/plugin/upstreamca/spire/spire.go

```diff
@@ -0,0 +1,184 @@
+	ServerAgentAddr string `hcl:"server_agent_address" json:"server_agent_address"`
+}
+
+type spirePlugin struct {
+	serialNumber      x509util.SerialNumber
+	workloadApiClient workload.X509Client
+
+	mtx               sync.RWMutex
+	cert              *x509.Certificate
+	upstreamCA        *x509svid.UpstreamCA
+
+	workloadKey       []byte
+	workloadCert      []byte
+	workloadBundle    []byte
+
+	spiffeID          string
+	trustDomain       url.URL
+	privateKey        *ecdsa.PrivateKey
+	intermediateCA    []*x509.Certificate
+	bundle            *bundleutil.Bundle
+
+	conn              *grpc.ClientConn
+	config 		  *Configuration
+}
+
+func (m *spirePlugin) Configure(ctx context.Context, req *plugin.ConfigureRequest) (*plugin.ConfigureResponse, error) {
+	// Parse HCL config payload into config struct
+	config := Configuration{}
+	if err := hcl.Decode(&config, req.Configuration); err != nil {
+		return nil, err
+	}
+
+
+	// fetch certificates and private key to be used for communication to node api
+	if req.GlobalConfig == nil {
+		return nil, errors.New("global configuration is required")
+	}
+
+	if req.GlobalConfig.TrustDomain == "" {
+		return nil, errors.New("trust_domain is required")
+	}
+
+	fmt.Println("[plugin][upstreamca] configured values [%v]",config)
+	fmt.Println("[plugin][upstreamca] global config [%v]",req.GlobalConfig)
+	fmt.Println("[plugin][upstreamca] global config trust domain [%v]",req.GlobalConfig.TrustDomain)
+
+	if req.GlobalConfig.TrustDomain != "" {
+		trustDomain, err := idutil.ParseSpiffeID("spiffe://" + req.GlobalConfig.TrustDomain, idutil.AllowAnyTrustDomain())
+		if err != nil {
+			return nil, err
+		}
+		m.trustDomain = *trustDomain
+	}
+
+	// create client for workload api
+	// get certificates and key
+	// set in the config
+	// do we need to continuously monitor the certificate for rotation ??
+	// the format for key , cert and bundle is ASN1 der format
+
+	m.config = &config
+
+	// TBD : need to move this submitCSR()
+	fmt.Println(" get workload certificates")
+	var err error
+	//m.workloadKey, m.workloadCert, m.workloadBundle, m.spiffeID, err = m.getWorkloadSVID(ctx, &config)
+	err = m.getWorkloadSVID(ctx, &config)
+	if err != nil {
+		fmt.Errorf("[plugin][upstreamca] failed to get workload certificates with err [%v]",err)
+		return nil, err
+	}
+
+	//TBD : need to remove the following calls , used for debugging
+
+	if m.workloadKey == nil {
+		fmt.Errorf("[plugin][upstreamca] invalid workload key ")
+		return nil, errors.New("invalid key format")
+	}
+
+	tmpkey, err := x509.ParsePKCS8PrivateKey(m.workloadKey)
+	if err != nil {
+		fmt.Errorf("[plugin][upstreamca] failed to parse workload key with err [%v]",err)
+		return nil, err
+	}
+	fmt.Println(" public key [%v]",tmpkey)
+
+	if m.workloadCert == nil {
+		fmt.Errorf("[plugin][upstreamca] invalid workload certificate")
+		return nil, errors.New("invalid workload certificate format")
+	}
+	//
+	_, err = x509.ParseCertificates(m.workloadCert)
```

Comment:
As noted in the comment above, this logic is best called when SubmitCSR is invoked. One thing to keep in mind is that this certificate will be rotating so it's best to fetch a new one when we need to use it

Reply:
Changes are added in SubmitCSR to query for workload certificates every time there is a request for spire core.